src/modules/adaptors/Morpho/MorphoAaveV3ATokenP2PAdaptor.sol

```diff
@@ -0,0 +1,154 @@
+    /**
+     * @dev Identifier unique to this adaptor for a shared registry.
+     * Normally the identifier would just be the address of this contract, but this
+     * Identifier is needed during Cellar Delegate Call Operations, so getting the address
+     * of the adaptor is more difficult.
+     */
+    function identifier() public pure override returns (bytes32) {
+        return keccak256(abi.encode("Morpho Aave V3 aToken P2P Adaptor V 1.1"));
+    }
+
+    /**
+     * @notice The Morpho Aave V3 contract on Ethereum Mainnet.
+     */
+    function morpho() internal pure returns (IMorpho) {
+        return IMorpho(0x33333aea097c193e66081E930c33020272b33333);
+    }
+
+    //============================================ Implement Base Functions ===========================================
+    /**
+     * @notice Cellar must approve Morpho to spend its assets, then call supply to lend its assets.
+     * @param assets the amount of assets to lend on Morpho
+     * @param adaptorData adaptor data containining the abi encoded aToken
+     * @param configurationData abi encoded maxIterations
+     */
+    function deposit(uint256 assets, bytes memory adaptorData, bytes memory configurationData) public override {
+        // Deposit assets to Morpho.
+        ERC20 underlying = abi.decode(adaptorData, (ERC20));
+        underlying.safeApprove(address(morpho()), assets);
+
+        uint256 iterations = abi.decode(configurationData, (uint256));
+        if (iterations == 0 || iterations > MAX_ITERATIONS) iterations = OPTIMAL_ITERATIONS;
+        morpho().supply(address(underlying), assets, address(this), iterations);
+
+        // Zero out approvals if necessary.
+        _revokeExternalApproval(underlying, address(morpho()));
+    }
+
+    /**
+     @notice Allows cellars to withdraw Morpho.
+     * @dev Important to verify that external receivers are allowed if receiver is not Cellar address.
+     * @param assets the amount of assets to withdraw from Morpho
+     * @param receiver the address to send withdrawn assets to
+     * @param adaptorData adaptor data containining the abi encoded ERC20 token
+     * @param configurationData abi encoded maximum iterations.
+     */
+    function withdraw(
+        uint256 assets,
+        address receiver,
+        bytes memory adaptorData,
+        bytes memory configurationData
+    ) public override {
+        // Run external receiver check.
+        _externalReceiverCheck(receiver);
+
+        ERC20 underlying = abi.decode(adaptorData, (ERC20));
+        uint256 iterations = abi.decode(configurationData, (uint256));
+
+        // Withdraw assets from Morpho.
+        morpho().withdraw(address(underlying), assets, address(this), receiver, iterations);
+    }
+
+    /**
+     * @notice Returns the p2p balance of the cellar.
+     */
+    function withdrawableFrom(bytes memory adaptorData, bytes memory) public view override returns (uint256) {
```

Comment:
I think we discussed this on our call, but can elaborate more if you want. 

I'll check and see if there is an easy way to figure out the available assets to withdraw based off the amount of assets supplied.

Reply:
Technically, balanceOf tells the cellar how much money it has there, withdrawableFrom tells the cellar how much money it CAN withdraw from the position. It is entirely possible for the 2 values to be the same.